src/test/socket/connect/test_connect.rs

```diff
@@ -0,0 +1,526 @@
+    let run_only_passing_tests = std::env::args().any(|x| x == "--shadow-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let tests = if run_only_passing_tests {
+        get_passing_tests()
+    } else {
+        get_all_tests()
+    };
+
+    test_utils::run_tests(&tests, summarize)?;
+
+    println!("Success.");
+    Ok(())
+}
+
+fn get_passing_tests() -> std::collections::BTreeMap<String, test_utils::TestFn> {
+    #[rustfmt::skip]
+    let mut tests: Vec<(String, test_utils::TestFn)> = vec![
+        ("test_invalid_fd".to_string(),
+            Box::new(test_invalid_fd)),
+        ("test_non_existent_fd".to_string(),
+            Box::new(test_non_existent_fd)),
+        ("test_null_addr".to_string(),
+            Box::new(test_null_addr)),
+        ("test_short_len".to_string(),
+            Box::new(test_short_len)),
+        ("test_zero_len".to_string(),
+            Box::new(test_zero_len)),
+    ];
+
+    // tests to repeat for different socket options
+    for &sock_type in [libc::SOCK_STREAM, libc::SOCK_DGRAM].iter() {
+        for &flag in [0, libc::SOCK_NONBLOCK, libc::SOCK_CLOEXEC].iter() {
+            // add details to the test names to avoid duplicates
+            let append_args = |s| format!("{} <type={},flag={}>", s, sock_type, flag);
+
+            #[rustfmt::skip]
+            let more_tests: Vec<(String, test_utils::TestFn)> = vec![
+                (append_args("test_after_close"),
+                    Box::new(move || test_after_close(sock_type, flag))),
+                (append_args("test_interface_loopback"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_LOOPBACK, None))),
+                (append_args("test_interface_loopback_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_LOOPBACK, Some(libc::INADDR_ANY)))),
+                (append_args("test_interface_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_ANY, None))),
+                (append_args("test_interface_any_any"),
+                    Box::new(move || test_interface(sock_type, flag, libc::INADDR_ANY, Some(libc::INADDR_ANY)))),
+                (append_args("test_double_connect_same_addr"),
+                    Box::new(move || test_double_connect(sock_type, flag, false))),
+                (append_args("test_double_connect_different_addr"),
+                    Box::new(move || test_double_connect(sock_type, flag, true))),
+            ];
+
+            tests.extend(more_tests);
+        }
+    }
+
+    let num_tests = tests.len();
+    let tests: std::collections::BTreeMap<_, _> = tests.into_iter().collect();
+
+    // make sure we didn't have any duplicate tests
+    assert_eq!(num_tests, tests.len());
+
+    tests
+}
+
+fn get_all_tests() -> std::collections::BTreeMap<String, test_utils::TestFn> {
```

Comment:
OK for this PR since it's the established pattern so far, but I wonder if it'd be a bit easier to read & maintain by adding a third parameter to each test, rather than having two lists. e.g.

```
let tests: Vec<(String, test_utils::TestFn, /* passing: */)bool> = vec![
    (append_args("test_after_close"),
      Box::new(move || test_after_close(sock_type, flag)),
      true),
    ...
     (append_args("test_non_existent_server"),
      Box::new(move || test_non_existent_server(sock_type, flag),
       false))
),
```

We could then construct the list of tests to actually run with a simple filter. That could be done from out here, though since all of our tests follow this pattern it could perhaps alternatively be moved into the test utility.

Reply:
Yeah I haven't really been happy with how the testing code is structured, so I'll try something like that outside of this PR.